flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/consumer/SingleInputGate.java

```diff
@@ -184,6 +184,9 @@
 	/** A timer to retrigger local partition requests. Only initialized if actually needed. */
 	private Timer retriggerLocalRequestTimer;
 
+	/** Flag indicating whether there is available data. */
+	private volatile boolean moreDataAvailable = false;
```

Comment:
(**hot looping without synchronisation part 1**) You should use `BufferOrEvent#moreAvailable()` to get next buffer and availability in the same call. If there is more data available, just keep reading from the gate without any extra synchronisation/callbacks. There is no need for `volatile boolean moreAvailable`. We always should use either this `BufferOrEvent#moreAvailable()` or relay on listeners/futures to inform us that the availability has changed.

Reply:
> We always should use either this BufferOrEvent#moreAvailable() or relay on listeners/futures to inform us that the availability has changed

I don't think they can accurately determine the state (available or unavailable). Considering the following execution sequences, the task thread finally sees the same thing that `BufferOrEvent#moreAvailable`  is unavailable and receives a non-null notification, but the last actual state of `InputGate` is different, the former is available, and the latter is unavailable.

Case 1:
1. Initial State: `InputGate` is data available
2. (**Task Thread**) Call `InputGate#pollNextBufferOrEvent()`
3. (**Task Thread**)  `InputGate` become empty after being polled a `BufferOrEvent`
4. (**Task Thread**)  Assign `BufferOrEvent#moreAvailable` to false before returning
5. (**Network Thread**) Data arrives at `InputGate`  which triggers the non-empty notification
6. (**Task Thread**) `InputGate#pollNextBufferOrEvent()` returns, `BufferOrEvent#moreAvailable` is false and recieves a `non-empty` notification

Case 2:
1. Initial State: `InputGate` is data unavailable
2. (**Task Thread**) Call `InputGate#pollNextBufferOrEvent()`
3. (**Network Thread**) Data arrives at `InputGate`  which triggers the non-empty notification
4. (**Task Thread**)  `InputGate` become empty after being polled a `BufferOrEvent`
5. (**Task Thread**)  Assign `BufferOrEvent#moreAvailable` to false before returning
6. (**Task Thread**) `InputGate#pollNextBufferOrEvent()` returns, `BufferOrEvent#moreAvailable` is false and recieves a `non-empty` notification

Reply:
Case 1: 
7. (Task Thread) received `BufferOrEvent#moreAvailable == false`, so it waits/checks on the `listenFutures[x]`. If there is any data, future should be completed.

Case 2:

I don't see a problem? If we get from notifications/futures false positive, what's the problem?